daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SExpr.scala

```diff
@@ -93,88 +95,75 @@ object SExpr {
   }
 
   object SEValue extends SValueContainer[SEValue]
 
-  /** Function application:
-    General case: 'fun' and 'args' are any kind of expression */
+  /** Function application: General case: 'fun' and 'args' are any kind of expression */
   final case class SEAppGeneral(fun: SExpr, args: Array[SExpr]) extends SExpr with SomeArrayEquals {
     def execute(machine: Machine): Unit = {
-      machine.pushKont(KArg(args, machine.frame, machine.actuals, machine.env.size))
-      machine.ctrl = fun
+      //machine.pushKont(KArg(args, machine.frame, machine.actuals, machine.env.size))
+      //machine.ctrl = fun
+      // We never encounter a general application node when executing on the speedy machine
+      // This is because all speedy expression must have been converted to ANF form
+      // The wrapper type `AExpr` helps ensure this restrction is adhered to.
+      // Better still would be to properly separate the `SExpr` and `AExpr` types.
+      throw SErrorCrash(s"execute: unexpected SEAppGeneral")
+    }
+  }
+
+  object SEApp {
+    //TODO: SEApp is no longer a smart constructor; remove this indirection.
+    def apply(fun: SExpr, args: Array[SExpr]): SExpr = {
+      SEAppGeneral(fun, args)
     }
   }
 
-  /** Function application:
-    Special case: 'fun' is an atomic expression. */
-  final case class SEAppAtomicFun(fun: SExprAtomic, args: Array[SExpr])
+  /** Function application: ANF case: 'fun' and 'args' are atomic expressions */
+  final case class SEAppAtomicGeneral(fun: SExprAtomic, args: Array[SExprAtomic])
       extends SExpr
       with SomeArrayEquals {
     def execute(machine: Machine): Unit = {
-      val vfun = fun.evaluate(machine)
-      executeApplication(machine, vfun, args)
+      val vfun = fun.lookupValue(machine)
+      //TODO: evaluate args to SValue here
+      enterApplication(machine, vfun, args)
     }
   }
 
-  /** Function application:
-    Special case: 'fun' is a builtin; size of `args' matches the builtin arity.
-    */
-  // A fully saturated builtin application
-  final case class SEAppSaturatedBuiltinFun(builtin: SBuiltin, args: Array[SExpr])
+  /** Function application: ANF case: 'fun' is builtin; 'args' are atomic expressions.  Size
+    * of `args' matches the builtin arity. */
+  final case class SEAppAtomicSaturatedBuiltin(
+      builtin: SBuiltinMaybeHungry,
+      args: Array[SExprAtomic])
       extends SExpr
       with SomeArrayEquals {
-    if (args.size != builtin.arity) {
-      throw SErrorCrash(s"SEAppB: arg.size != builtin.arity")
-    }
     def execute(machine: Machine): Unit = {
       val arity = builtin.arity
       val actuals = new util.ArrayList[SValue](arity)
-      machine.pushKont(KBuiltin(builtin, actuals, machine.env.size))
-      evaluateArguments(machine, actuals, args, args.length);
+      for (i <- 0 to arity - 1) {
+        val arg = args(i)
+        val v = arg.lookupValue(machine)
+        actuals.add(v)
+      }
+      builtin.execute(actuals, machine)
```

Comment:
Once this PR has landed, let's please benchmark if passing the `args` directly to `builtin.execute` and letting the latter do the lookups makes a significant difference. This should avoid the allocation of lots of very short-lived arrays. This applies in quite a few other places too.

Reply:
I made a issue for this: #6522

Reply:
Merci.